src/arch/sparc64/CodeGen.zig

```diff
@@ -1884,6 +2042,101 @@ fn airRetPtr(self: *Self, inst: Air.Inst.Index) !void {
     return self.finishAir(inst, .{ .ptr_stack_offset = stack_offset }, .{ .none, .none, .none });
 }
 
+fn airShlSat(self: *Self, inst: Air.Inst.Index) !void {
+    const bin_op = self.air.instructions.items(.data)[inst].bin_op;
+    const result: MCValue = if (self.liveness.isUnused(inst)) .dead else return self.fail("TODO implement shl_sat for {}", .{self.target.cpu.arch});
+    return self.finishAir(inst, result, .{ bin_op.lhs, bin_op.rhs, .none });
+}
+
+fn airShlWithOverflow(self: *Self, inst: Air.Inst.Index) !void {
+    const ty_pl = self.air.instructions.items(.data)[inst].ty_pl;
+    const extra = self.air.extraData(Air.Bin, ty_pl.payload).data;
+    const result: MCValue = if (self.liveness.isUnused(inst)) .dead else result: {
+        const lhs = try self.resolveInst(extra.lhs);
+        const rhs = try self.resolveInst(extra.rhs);
+        const lhs_ty = self.air.typeOf(extra.lhs);
+        const rhs_ty = self.air.typeOf(extra.rhs);
+
+        switch (lhs_ty.zigTypeTag()) {
+            .Vector => return self.fail("TODO implement mul_with_overflow for vectors", .{}),
+            .Int => {
+                const int_info = lhs_ty.intInfo(self.target.*);
+                if (int_info.bits <= 64) {
+                    try self.spillConditionFlagsIfOccupied();
+
+                    const lhs_lock: ?RegisterLock = if (lhs == .register)
+                        self.register_manager.lockRegAssumeUnused(lhs.register)
+                    else
+                        null;
+                    // TODO this currently crashes stage1
```

Comment:
Yep, it's fine provided we remember to change it back to `defer`s once stage2 is out. Also, note that I have been battling with this myself when rewriting the register freezing logic, and in general a crash in stage1 is due to block nesting and having defers in deeply nested block, and/or problems with automatic type inference. For example, in your case, if you elide the outer-most block by early returning when liveness suggests instruction is dead like so

```zig
if (self.liveness.isUnused(inst)) {
    return self.finishAir(inst, .dead, .{ extra.lhs, extra.rhs, .none });
 }
const lhs = try self.resolveInst(extra.lhs);
```

I believe this will be allow you to put `defer`s back in.

Reply:
Yeah, I'm happy to leave it in with the promise of fixing it once stage2 is released :-)